x/crosschain/keeper/msg_server_vote_outbound.go

```diff
@@ -136,18 +138,105 @@ func (k msgServer) VoteOutbound(
 	return &types.MsgVoteOutboundResponse{}, nil
 }
 
-// FundStabilityPool funds the stability pool with the remaining fees of an outbound tx
-// The funds are sent to the gas stability pool associated with the receiver chain
-// This wraps the FundGasStabilityPoolFromRemainingFees function and logs an error if it fails.We do not return an error here.
+// ManageUnusedGasFee funds the stability pool with the remaining fees of an outbound tx
+// The funds are sent to the gas stability pool associated with the receiver chain and the user is refunded if possible.
+// This wraps the UseRemainingGasFee function and logs an error if it fails.We do not return an error here.
 // Event if the funding fails, the outbound tx is still processed.
-func (k Keeper) FundStabilityPool(ctx sdk.Context, cctx *types.CrossChainTx) {
-	// Fund the gas stability pool with the remaining funds
-	if err := k.FundGasStabilityPoolFromRemainingFees(ctx, *cctx.GetCurrentOutboundParam(), cctx.GetCurrentOutboundParam().ReceiverChainId); err != nil {
-		ctx.Logger().
-			Error("%s: CCTX: %s Can't fund the gas stability pool with remaining fees %s", voteOutboundID, cctx.Index, err.Error())
+func (k Keeper) ManageUnusedGasFee(ctx sdk.Context, cctx *types.CrossChainTx) {
+	outboundParams := cctx.GetCurrentOutboundParam()
+	// If the userGasFeePaid is nil, it means that this cctx was created before the userGasFeePaid was added.
+	if outboundParams.UserGasFeePaid.IsNil() {
+		err := k.FundGasStabilityPoolFromRemainingFees(
+			ctx,
+			*outboundParams,
+			outboundParams.ReceiverChainId,
+		)
+		if err != nil {
+			ctx.Logger().Error("Failed to fund gas stability pool (Legacy) with remaining fees",
+				"voteOutboundID", voteOutboundID,
+				"cctxIndex", cctx.Index,
+				"error", err,
+			)
+		}
+		return
+	}
+
+	if err := k.useRemainingGasFee(ctx, cctx); err != nil {
+		ctx.Logger().Error("Failed to fund gas stability pool with remaining fees",
+			"voteOutboundID", voteOutboundID,
+			"cctxIndex", cctx.Index,
+			"error", err,
+		)
+	}
+}
+
+// UseRemainingGasFee uses the remaining fees of an outbound tx to fund the gas stability pool and refund the user if possible
+func (k Keeper) useRemainingGasFee(
+	ctx sdk.Context,
+	cctx *types.CrossChainTx,
+) error {
+	OutboundParams := cctx.GetCurrentOutboundParam()
+
+	receiverChainID := OutboundParams.ReceiverChainId
```

Comment:
up to you, but i think visually this function can be a bit smaller if there are no variables like this one

also i would move validation like this at beginning so there is no need to any calculation if can fail fast:
```
chainParams, found := k.GetObserverKeeper().GetChainParamsByChainID(ctx, receiverChainID)
	if !found {
		return errors.Wrap(observertypes.ErrChainParamsNotFound, fmt.Sprintf("chainID: %d", receiverChainID))
	}
```